File: Java Programs/BinarySearch.java
import lib.SingleArray;
/*
Program to perform Binary Search on an Array using ADT to store and retrieve Elements.
1. Array should be in the Sorted Fashion
2. Time Complexity is 
   Best Case - O(1)
   Worst Case - O(logn)
   Average Case - O(logn)
*/
public class BinarySearch {
    public static void main(String[] args) {
        //Instantiation of Class
        SingleArray myArray = new SingleArray(10);
        //Set Array Elements
        myArray.setData(5);
        myArray.setData(9);
        myArray.setData(11);
        myArray.setData(17);
        myArray.setData(20);
        myArray.setData(45);
        myArray.setData(78);
        myArray.setData(98);
        //Display array Data
        myArray.showArray();
        //Instantiating Parent Class
        BinarySearch bs = new BinarySearch();
        //Calling Binary Search on Array
        int foundPos = bs.binarySearch(myArray,46);
        if (foundPos!=-1)
        System.out.println("Element Found at:"+foundPos);
        else
        System.out.println("Element Not Found.");
    }
   //Method to search an element in a given array
   int binarySearch(SingleArray myArray, int data){
        //Logic goes based on low high mid indices. It starts searching with mid and tries to find if the number is greater or lesser than mid
        //accordingly new low high mid values will be set. This goes on either until the element is found or until the low > high
        int low,high,mid;
        low = 0;
        high = myArray.getUsedSize();
        
        while(low<=high){
            mid = (low+high)/2;
            if(data<myArray.getData()[mid]){
                high = mid-1;
            }else if (data > myArray.getData()[mid]){
                low = mid+1;
            }else{
               return mid+1;
            }
        }
        return -1;
    }
}

File: C Programs/BinarySearch.c
#include<stdio.h>
#include<malloc.h>
/*Program to perform Binary Search on an Array using ADT to store and retrieve Elements.
1. Array should be in the Sorted Fashion
2. Time Complexity is 
   Best Case - O(1)
   Worst Case - O(logn)
   Average Case - O(logn)
*/
struct myArray{
    int total_size;
    int used_size;
    int *ptr;
};

void createArray(struct myArray *arr, int tSize){
    arr->total_size = tSize;
    arr->used_size = 0;
    arr->ptr = (int *)malloc(tSize*sizeof(int));
}

void show(struct myArray *a){
  for(int i=0; i< a->used_size; i++){
    printf("%d\n",(a->ptr)[i]);
  }
}

void setVal(struct myArray *arr, int position, int data){
    (arr->ptr)[position] = data;
    arr->used_size = arr->used_size+1;
}

/*
Logic goes based on low high mid indices. It starts searching with mid and tries to find if the number is greater or lesser than mid
accordingly new low high mid values will be set. This goes on either until the element is found or until the low > high
*/
int binarySearch(struct myArray *arr, int data){
    int low, mid, high;
    high = (arr->used_size);
    low = 0;
    mid = high/2;
    for(int i = 0; i<(arr->used_size)-1; i++){
        mid = (low+high)/2;
        printf("Number of Seaches %d\n",i+1);
        if(low>high){
            printf("Element Not Found!");
            break;
        }
       if ((arr->ptr)[mid] > data){
           high = mid-1;
       }else if  ((arr->ptr)[mid] < data){
           low = mid+1;
       }else{
           printf("Element Found at %d Position",mid+1);
           break;
       }
    }
    
}
// Binary Searching for Sorted Arrays.
void main(){
    struct myArray arrData;
    //Create an Array
    createArray(&arrData, 10 );
    //Set the data to an Array
    setVal(&arrData,0, 5);
    setVal(&arrData,1, 9);
    setVal(&arrData,2, 11);
    setVal(&arrData,3, 17);
    setVal(&arrData,4, 20);
    setVal(&arrData,5, 45);
    setVal(&arrData,6, 78);
    setVal(&arrData,7, 99);
    //Display array Data
    show(&arrData);
    //Calling Binary Search on Array
    binarySearch(&arrData,46);
}